src/nrnoc/vclmp.mod

```diff
@@ -99,7 +99,7 @@ INITIAL {
 		: stopping cvode here if the clamp is on still allows
 		: that tool to control a IClamp under cvode
 		VERBATIM
-		{extern int cvode_active_;
+		{
```

Comment:
In general it seems preferable to stop wrapping translated code with `extern "C" { ... }`; I think removing that would make this change unnecessary, but there are various other wrinkles to iron out...

Reply:
I don't see any legacy mod files in modeldb that contain ```extern "C"``` but of course the effective NEURON API is/was, to a large extent, ```extern "C"```. There are very many legacy ```VERBATIM``` statements that have ```extern ...``` statements instead of ```#include```
Is it possible to make the entire body of the generated c++ file surrounded by ```extern "C" {```? Or does that defeat your purpose?

Reply:
`nocmodl` already wraps the bulk of the C/C++ code it generates in `extern "C"`, it seems (adapted from `
src/nrnoc/hh.cpp`):
```c++
#include ...

/* ... */

#define ...

/* ... */

#if defined(__cplusplus)
extern "C" {
#endif

/* ... all function declarations and definitions ... */

#if defined(__cplusplus)
} /* extern "C" */
#endif

/* ... a little bit more NMODL_TEXT stuff */
```

There here been a few different issues with compiling this as C++ instead of C.

One is that while various mod files declare NEURON methods they want to call, they only specify the name and not the prototype, for example: https://github.com/neuronsimulator/nrn/blob/edb982ebb9913bcd18461f5d5b119c6215280bb7/src/nrnoc/apcount.mod#L22-L26
and then call these functions using some approximation to the real prototype, for example:
https://github.com/neuronsimulator/nrn/blob/edb982ebb9913bcd18461f5d5b119c6215280bb7/src/nrnoc/apcount.mod#L32-L35
which is valid C, but not valid C++. I say "some approximation to" because, again for the example above, the real definition
https://github.com/neuronsimulator/nrn/blob/edb982ebb9913bcd18461f5d5b119c6215280bb7/src/ivoc/ivocvect.cpp#L269
takes a `IvocVect*`, not a `void*`. To me the natural solution here is to suppress the declarations in the translated code and to make sure the relevant header is included. I already did that for these `vector_` methods in this PR.

A second part of this is that C++ does not allow passing a `void*` to a function expecting some other pointer type. This means that code like this:
https://github.com/neuronsimulator/nrn/blob/edb982ebb9913bcd18461f5d5b119c6215280bb7/src/nrnoc/pattern.mod#L138-L140
is problematic when compiled as C++, assuming one does not want to abandon type safety by making the only "real" prototype of `vector_capacity` take `void*` (and assuming one does not want to play games with `extern "C"` and different prototypes in different places). My proposed solution in this PR has two parts:
- give the `vector_*` methods C++ linkage (i.e. drop `extern "C"`), so overloading is available
- add type-unsafe overloads with `void*` arguments, and mark these as deprecated except when compiling legacy models

unfortunately this is less clean for `vector_new*`, because we cannot overload by return type, but it seems that there is already a family of old (`vector_new{1,2,3}`) and new (`vector_new`) methods, and there seems to be a combination of prototypes that works.

For other (not `vector_*`) methods I think we will have to see case-by-case. Leaving these `extern "C"` avoids some pedantry about having consistently declared/used prototypes, but also prevents us from using C++ features (such as overloading) to smooth over such inconsistencies. In the case of global variables (like the change that this discussion is commenting on) we see yet another related bit of pedantry along the same lines, which GCC tolerates but Clang does not: https://godbolt.org/z/4jn5G7T31. In this case I can probably also fix it without touching the `.mod` file by making the global explicitly `extern "C"`.